### PR TITLE
Refactor evm into new module

### PR DESCRIPTION
**Pull request by lightclient**
This refactors the evm logic into it's own package. I think this is a good move since it is quite separate from the general state transition logic.

I quite like how simple the instructions look with the helper functions `push, pop, subtract_gas`. We'll just need to have some exception handlers in the interpreter to handle those. @SamWilsn I know at first you didn't seem too excited about raising exceptions, how do you feel about it now?

### CODE REVIEW

File: src/eth1spec/trie.py [hunk 1]
```diff
@@ -220,6 +224,9 @@ def c(J: Mapping[Bytes, Union[Bytes, Account, Receipt]], i: Uint) -> Bytes:
     # if leaf node
     if len(J) == 1:
         leaf = J[I_0]
+
+        I_1: Union[bytes, Uint]
```
Comment (lightclient) [inline src/eth1spec/trie.py:228]: This seems a bit unfortunate and is a byproduct of having the storage trie be a `Dict[Bytes32, U256]`. I think when I rework this code I can improve some of these weird typings.

**Comment by lightclient** [comment]
closes #19

File: src/eth1spec/evm/error.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+"""
+EVM Runtime Errors
+---
```
Comment (SamWilsn) [inline src/eth1spec/evm/error.py]:
```suggestion
EVM Runtime Errors
------------------
```

Underlines have to match their headings (I think.)

File: src/eth1spec/evm/interpreter.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+A straightforward interpreter that executes EVM code.
+"""
+
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from ..eth_types import EMPTY_ACCOUNT, U256, Address, Hash32, Log, State, Uint
+from . import Environment, Evm
+from .gas import GAS_VERY_LOW
+from .ops import *  # noqa: F403
```
Comment (SamWilsn) [inline src/eth1spec/evm/interpreter.py]: This really limits the effectiveness of the linter, so we should avoid star imports.

File: src/eth1spec/evm/ops.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# flake8: noqa 405
+
+from .instructions import *  # noqa 403
```
Comment (SamWilsn) [inline src/eth1spec/evm/ops.py]: Same comment about star imports.

File: src/eth1spec/evm/ops.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# flake8: noqa 405
```
Comment (SamWilsn) [inline src/eth1spec/evm/ops.py]:
```suggestion
"""
Instruction Encoding (Opcodes)
------------------------------

Machine readable representations of EVM instructions, and a mapping to their implementations.
"""
```

**Review by SamWilsn — CHANGES_REQUESTED** [review]
Sorry to be so pedantic!

**Comment by SamWilsn** [comment]
I think the exceptions might be okay. I'm okay seeing where they lead.

File: src/eth1spec/evm/interpreter.py (hunk 3 above)
Comment (lightclient) [reply src/eth1spec/evm/interpreter.py]: Okay :(

**Comment by lightclient** [comment]
> Sorry to be so pedantic!

No no, this is the best time to be pedantic :)

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**